### PR TITLE
RM-8329 Simplify the implementation of AsyncLocalStorageProvider

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-8329

### CODE REVIEW

File: Remotion/Web/Core/Context/HttpContextStorageProvider.cs [hunk 1]
```diff
@@ -34,6 +34,11 @@ public class HttpContextStorageProvider : ISafeContextStorageProvider
 #pragma warning restore 618
 #endif
 
+    public SafeContextBoundary OpenSafeContextBoundary ()
+    {
+      return default;
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: I think we need the new semantic with the cleanup also for HttpContextStorageProvider in .NET 5. 

And since the semantics shouldn't differ between .net 5 and .net Framework, we also need this for Callcontext-Storageprovider.

**Review by MichaelKetting — COMMENTED** [review]
Partial review due to the notes for SafeContextBoundary

File: Remotion/Web/Core/Context/HttpContextStorageProvider.cs (hunk 1 above)
Comment (a-ctor) [reply Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: `HttpContextStorageProvider` writes the values directly into `HttpContext.Items`. How should a reset behave in this case?

Comment (MichaelKetting) [reply Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: hmm....removing and re-adding while the scope is open. this will "guard" scenarios such as parallel.foreach. For task etc, httpcontext should not propagate to the new thread anyway.

Comment (a-ctor) [reply Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: So we need to keep track of all values written to be able to remove those values at the boundary and restore the old values afterwards? But what about values from the inside scope leaking out?

We would need to save the complete `Items` collection and set it back the way it was. Which is kind of awkward since we cannot set the `Items` directly, but have to make a diff and do `Add`/`Remove` calls

File: Remotion/Core/Core/Context/ISafeContextBoundaryStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Remotion.Context
+{
+  /// <summary>
+  /// Implements a strategy for <see cref="SafeContextBoundary"/> to restore a previous safe context.
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/ISafeContextBoundaryStrategy.cs:4]:
```suggestion
  /// Implements a strategy for <see cref="SafeContextBoundary"/> to restore a previous context.
```
It's a consistency thing. `SafeContext` as a standalone type isn't an exact match, but reading `safe context` as two words seems odd.

File: Remotion/Core/Core/Context/ISafeContextBoundaryStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  /// </summary>
+  /// <seealso cref="SafeContextBoundary"/>
+  /// <seealso cref="ISafeContextStorageProvider.OpenSafeContextBoundary"/>
+  public interface ISafeContextBoundaryStrategy
+  {
+    /// <summary>
+    /// Restores a previous state of the safe context from <paramref name="state"/>.
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/ISafeContextBoundaryStrategy.cs:11]:
```suggestion
    /// Restores a previous state of the context from <paramref name="state"/>.
```

File: Remotion/Core/Core/Context/ISafeContextStorageProvider.cs [hunk 4]
```diff
@@ -24,6 +24,13 @@ namespace Remotion.Context
   // By default, use HttpContextStorageProvider. Only if not available, use CallContextStorageProvider.
   public interface ISafeContextStorageProvider
   {
+    /// <summary>
+    /// Opens a <see cref="SafeContextBoundary"/> into which values will not flow.
+    /// The previous context is restored when the <see cref="SafeContextBoundary"/> is disposed.
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/ISafeContextStorageProvider.cs:29]:
```suggestion
    /// Opens a <see cref="SafeContextBoundary"/>, preventing the values in the current context from flowing to the new context.
    /// The previous context is restored when the <see cref="SafeContextBoundary"/> is disposed.
```

File: Remotion/Core/Core/Context/ISafeContextStorageProvider.cs [hunk 5]
```diff
@@ -24,6 +24,13 @@ namespace Remotion.Context
...
+    /// </summary>
+    /// <returns>A <see cref="SafeContextBoundary"/> object with which the previous context can be restored.</returns>
+    SafeContextBoundary OpenSafeContextBoundary ();
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/ISafeContextStorageProvider.cs:32]: The docs also imply that the boundary is unconditional, i.e. also works with CallContext and HttpContext.

File: Remotion/Core/Core/Context/SafeContextBoundary.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+namespace Remotion.Context
+{
+  /// <summary>
+  /// Represents a <see cref="SafeContext"/> boundary, across which values will not flow.
+  /// The previous context is restored when disposed.
+  /// </summary>
+  /// <seealso cref="ISafeContextStorageProvider.OpenSafeContextBoundary"/>
+  public readonly struct SafeContextBoundary : IDisposable
+  {
+    private readonly ISafeContextBoundaryStrategy _boundaryStrategy;
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/SafeContextBoundary.cs:12]: Why isn't NRT complaining that the `?` is missing?

File: Remotion/Core/UnitTests/Context/SafeContextBoundaryTests.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      var state = new object();
+
+      var safeContextBoundary = new SafeContextBoundary(boundaryStrategyMock.Object, state);
+      safeContextBoundary.Dispose();
+
+      boundaryStrategyMock.Verify(e => e.RestorePreviousSafeContext(state));
+    }
+
+    [Test]
+    public void Dispose_WithNullBoundaryStrategy_ThrowsNothing ()
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Context/SafeContextBoundaryTests.cs:23]:
```suggestion
    public void Dispose_WithDefaultInitialization_DoesNotThrow ()
```

File: Remotion/Core/UnitTests/Context/SafeContextBoundaryTests.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    {
+      var safeContextBoundary = new SafeContextBoundary();
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Context/SafeContextBoundaryTests.cs:25]: I wonder if we should use `default` here instead. C# 10 allows default ctors for structs with an implementation so this code might mask an issue in the future.

**Review by MichaelKetting — COMMENTED** [review]
partial review, AsyncLocal-impl is left to review

File: Remotion/Web/Core/Context/HttpContextStorageProvider.cs (hunk 1 above)
Comment (MichaelKetting) [reply Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: Actually: place a dictionary into the Items and switch the dictionary. Same as with AsyncLocal

Comment (MichaelKetting) [reply Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: (I know, it's different than before and should be documented, but over all, that solves things.)

File: Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs [hunk 9]
```diff
@@ -106,19 +74,71 @@ public void SetData_FromOtherThread_NotAccessible ()
...
+    }
+
+    [Test]
+    public void SetData_FromThreadWithSafeContextBoundary_DoesNotOverrideValueOnMainThread ()
+    {
+      _provider.SetData(c_testKey, "1");
+
+      var t = new Thread(() =>
+      {
+        using (_provider.OpenSafeContextBoundary())
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs:96]: We should also add a test that shows that disposing the boundary is optional

File: Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs [hunk 10]
```diff
@@ -17,234 +17,72 @@
...
-using Remotion.Logging;
 
 namespace Remotion.Context
 {
 #if !NETFRAMEWORK
   [ImplementationFor(typeof(ISafeContextStorageProvider), Lifetime = LifetimeKind.Singleton, Position = 1)]
 #endif
-  [Obsolete("This type is experimental and might have unintended side effects. (Version 3.0.0-alpha.22)")]
-  public class AsyncLocalStorageProvider : ISafeContextStorageProvider
+  public class AsyncLocalStorageProvider : ISafeContextStorageProvider, ISafeContextBoundaryStrategy
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs:30]: XML Doc

File: Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs [hunk 11]
```diff
@@ -17,234 +17,72 @@
...
-    }
-
-    private AsyncLocalStorageProvider (bool tracingEnabled)
     {
-      _tracingEnabled = tracingEnabled;
-      _context = new AsyncLocal<StorageContext>(AsyncLocalValueChangedHandler);
     }
 
-    private void AsyncLocalValueChangedHandler (AsyncLocalValueChangedArgs<StorageContext> args)
+    public SafeContextBoundary OpenSafeContextBoundary ()
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs:38]: inherit doc?

File: Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs [hunk 12]
```diff
@@ -17,234 +17,72 @@
...
+      _context.Value?.Remove(key);
+    }
 
-      var newValue = oldValue.FreeData(key);
-      if (oldValue != newValue)
-        _context.Value = newValue;
+    /// <inheritdoc />
+    void ISafeContextBoundaryStrategy.RestorePreviousSafeContext (object? state)
+    {
+      _context.Value = state as Hashtable;
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs:85]: Hard cast.

Soft cast is only used when a type test is explicitly intended and a mismatched type does not indicate a bug. Here, an invalid state type will mean something's gone horribly wrong.

File: Remotion/Web/Core/Context/HttpContextStorageProvider.cs (hunk 1 above)
Comment (MichaelKetting) [reply Remotion/Web/Core/Context/HttpContextStorageProvider.cs:39]: Performance question: 
1. the lookup speed will depend on the key-length, the key should be just a few characters (see SafeContextKeys).
2. how does the performance of a weakreference-valueget compare to the lookup in the HttpContext.Items? If the weakreference is significantly better, we could discuss a performance optimization by caching the last dictonary in a threadlocal weakreference (there's a bit more involved...)

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Core/Core/Context/SafeContextBoundary.cs (hunk 6 above)
Comment (a-ctor) [reply Remotion/Core/Core/Context/SafeContextBoundary.cs:12]: Struct values could always be `null`, but that would not make sense with NRT. So, it is possible to define values that are non-nullable and the null checks guard only the assignment part in the constructor.

File: Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs [hunk 13]
```diff
@@ -274,97 +455,226 @@ public void SuppressFlowWorksCorrectly ()
...
+    public void SetData_InParallelForeachWithUnclosedSafeContextBoundary_DoesNotSetValueForOuterScope ()
+    {
+      _provider.SetData(c_testKey, "initial");
+
+      Parallel.For(
+          0,
+          100,
+          v =>
+          {
+            _provider.OpenSafeContextBoundary();
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs:670]: this works, too? great!

**Review by MichaelKetting — APPROVED** [review]
please don't forget the httpcontext followup ticket
